### PR TITLE
bpo-38525: Fix a segmentation fault when using reverse iterators of empty dict

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue38525

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2019-10-19-23-58-09.bpo-38525.CzEl2G.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix dictreviter_iternext for empty :class:`dict` Patch by Dong-hee Na.
```
Comment (pablogsal) [inline Misc/NEWS.d/next/Library/2019-10-19-23-58-09.bpo-38525.CzEl2G.rst]:
```suggestion
Fix a segmentation fault when using reverse iterators of empty ``dict``
objects that were created as empty ``dict`` literals.  Patch by Dong-hee Na.
```

File: Lib/test/test_dict.py [hunk 2]
```diff
@@ -1312,6 +1312,12 @@ def test_reversed(self):
         self.assertEqual(list(r), list('dcba'))
         self.assertRaises(StopIteration, next, r)
 
+    def test_empty_reversed(self):
```
Comment (pablogsal) [inline Lib/test/test_dict.py]: Can you add the same test for the non-literal version and for the rest of the iterators:

```
        the_dict = {}
        self.assertEqual(list(reversed(the_dict)), list())
        self.assertEqual(list(reversed(the_dict.values())), list())
        self.assertEqual(list(reversed(the_dict.keys())), list())

        the_dict = dict()
        self.assertEqual(list(reversed(the_dict)), list())
        self.assertEqual(list(reversed(the_dict.values())), list())
        self.assertEqual(list(reversed(the_dict.keys())), list())
```

Comment (pablogsal) [inline Lib/test/test_dict.py]:
```suggestion
    def test_reverse_iterator_for_empty_dict(self):
```

**Comment by pablogsal** [comment]
CC: @remilapeyre

**Review by pablogsal — CHANGES_REQUESTED** [review]
Good work! I have left some minor comments and suggestions

**Comment by corona10** [comment]
@pablogsal 

Thanks for the suggestion!
I've updated the PR :)

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/test/test_dict.py [hunk 3]
```diff
@@ -1312,6 +1312,19 @@ def test_reversed(self):
         self.assertEqual(list(r), list('dcba'))
         self.assertRaises(StopIteration, next, r)
 
+    def test_reverse_iterator_for_empty_dict(self):
+        empty_dict = {}
+        self.assertEqual(list(reversed(empty_dict)), [])
```
Comment (vstinner) [inline Lib/test/test_dict.py]: This test may not reproduce the reference count crash. You should remove empty_dict and write {} directly. But please add a comment to explain why you do that.

File: Lib/test/test_dict.py [hunk 4]
```diff
@@ -1312,6 +1312,19 @@ def test_reversed(self):
         self.assertEqual(list(r), list('dcba'))
         self.assertRaises(StopIteration, next, r)
 
+    def test_reverse_iterator_for_empty_dict(self):
```
Comment (vstinner) [inline Lib/test/test_dict.py:1315]: Please add a reference to the bpo in a comment and a short explanation of the crash.

File: Lib/test/test_dict.py [hunk 5]
```diff
@@ -1312,6 +1312,19 @@ def test_reversed(self):
...
+        empty_dict = {}
+        self.assertEqual(list(reversed(empty_dict)), [])
+        self.assertEqual(list(reversed(empty_dict.items())), [])
+        self.assertEqual(list(reversed(empty_dict.values())), [])
+        self.assertEqual(list(reversed(empty_dict.keys())), [])
+
+        empty_dict = dict()
```
Comment (vstinner) [inline Lib/test/test_dict.py]: These tests look redundant.

Comment (pablogsal) [reply Lib/test/test_dict.py]: They are not redudant because `dict` and `{}` don't trigger the same code path.

Indeed, the error is not reproducible by

`list(reversed(dict()))`

but only for

`list(reversed({}))`

So I think it makes sense to test both code paths for more coverage.

**Review by rhettinger — APPROVED** [review]

File: Lib/test/test_dict.py (hunk 5 above)
Comment (corona10) [reply Lib/test/test_dict.py]: @pablogsal is correct. I will re-add dict() case.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
There is other problem here. See bpo for details.

File: Misc/NEWS.d/next/Core and Builtins/2019-10-20-00-36-18.bpo-38525.Vty1cA.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Fix a segmentation fault when using reverse iterators of empty ``dict``
+objects that were created as empty ``dict`` literals. Patch by Dong-hee Na.
```
Comment (methane) [inline Misc/NEWS.d/next/Core and Builtins/2019-10-20-00-36-18.bpo-38525.Vty1cA.rst]: It is implementation detail and not correct (e.g. `d.clear()` makes key-sharing empty dict too).
I think this is enough:

```
Fix a segmentation fault when using reverse iterators of empty dict objects.
```

Comment (corona10) [reply Misc/NEWS.d/next/Core and Builtins/2019-10-20-00-36-18.bpo-38525.Vty1cA.rst]: @methane Thanks I've updated it :)

**Comment by remilapeyre** [comment]
Sorry for botching that, I will look into the issue with shared keys

File: Misc/NEWS.d/next/Core and Builtins/2019-10-20-00-36-18.bpo-38525.Vty1cA.rst [hunk 7]
```diff
@@ -0,0 +1,2 @@
+Fix a segmentation fault when using reverse iterators of empty ``dict`` objects.
+Patch by Dong-hee Na and Inada Naoki.
```
Comment (corona10) [inline Misc/NEWS.d/next/Core and Builtins/2019-10-20-00-36-18.bpo-38525.Vty1cA.rst:2]: @methane 
I 've added your name to NEWS entry.

File: Objects/dictobject.c [hunk 8]
```diff
@@ -3826,7 +3831,7 @@ dictreviter_iternext(dictiterobject *di)
     PyObject *key, *value, *result;
 
     if (d->ma_values) {
-        if (i < 0) {
+        if (i < 0 || i >= d->ma_used) {
```
Comment (serhiy-storchaka) [inline Objects/dictobject.c]: Is this change still needed?

Comment (corona10) [reply Objects/dictobject.c]: @serhiy-storchaka 
No, it is not needed. I removed it on the latest commit.

**Review by serhiy-storchaka — APPROVED** [review]

**Review by pablogsal — APPROVED** [review]

**Comment by vstinner** [comment]
Nice contributions @corona10, it's rare to have to fix a bug in the dict implementation :-)
